### ISSUE TITLE
[Issue]: Huge `<title>` elements in `**/*.nfo` files DOS the DB and `RES` RAM sizes

### ISSUE THREAD

**Issue by rpatterson** [jellyfin/jellyfin#12232]
### Please describe your bug

TL;DR: Jellyfin "trusts" `**/*.nfo` XML content which can lead to an effective DOS by bloating the DB. Jellyfin should sanitize its inputs from `**/*.nfo` XML and truncate large values.

A library contained `**/*.nfo` files for most media files and a few of those files contained valid XML but bad content that bloated them to huge 100+ MB sizes. The text content of the `<title>` elements made up most of that size in those huge files.

Starting sometime around the `FixAudioData` migration:

```diff
CommitDate: Mon Jun 3 00:58:25 2024

./config/migrations.xml | 4 ++++

modified   ./config/migrations.xml
@@ -73,5 +73,9 @@
       <Item1>852816e0-2712-49a9-9240-c6fc5fcad1a8</Item1>
       <Item2>UpdateDefaultPluginRepository10.9</Item2>
     </ValueTupleOfGuidString>
+    <ValueTupleOfGuidString>
+      <Item1>cf6fabc2-9fbe-4933-84a5-ffe52ef22a58</Item1>
+      <Item2>FixAudioData</Item2>
+    </ValueTupleOfGuidString>
   </Applied>
 </MigrationOptions>
\ No newline at end of file
```

the `/config/data/library.db` grew to ~10x it's previous size as observed in the corresponding backup files:

```console
$ ls -ldht ./config/data/library.db.bak[123]
-rw-r--r-- 1 1000 users 4.3G Jun 30 15:11 ./config/data/library.db.bak3
-rw-r--r-- 1 1000 users 489M Jun  2 00:25 ./config/data/library.db.bak2
-rw-r--r-- 1 1000 users 490M May 12 11:16 ./config/data/library.db.bak1
```

The Jellyfin process startup began taking ~20 minutes before responding to browser requests and both library scans and Kodi library repairs would cause `RES` memory usage to jump and exhaust the physical RAM and often the process would die or be killed.

After cleaning up the content in those huge `**/*.nfo` files, restoring the `./config/data/library.db.bak2` DB backup, and removing the entry for the `FixAudioData` migration `./config/migrations.xml`, the Jellyfin process starts quickly and both Jellyfin library scans and Kodi library repairs complete without bloating the DB size or `RES` RAM usage.

This DB bloat seems to be a change in Jellyfin's behavior, because these malformed `**/*.nfo` files have been in the library for years, confirmed with `$ ls -l --time=creation`, so this may be some sort of regression. Even if the `**/*.nfo` files were damaged recently, this should still be considered a bug in Jellyfin, because it's vulnerable to malformed data from external sources and being defensive about that and sanitizing inputs is Jellyfin's responsibility even if it's *also* a bug in whatever generated the bad XML content. A few bad `**/*.nfo` files from public torrents, for example, should not be able to bring down the whole media server.

Most importantly, the component that ingests `**/*.nfo` XML should truncate strings longer than a global string length limit. Next most important would be to truncate lists and mappings longer than another global limit. I think it would also be good to audit which `**/*.nfo` XML "fields" end up getting indexed in the DB and truncating those values to a more defensive, shorter limit because long data that is indexed is likely to affect DB performance and truncating absurdly long data in the *indexes* isn't likely to be experienced as "lost data" by users. From the bad `**/*.nfo` XML I found in my library, it should be pretty easy to set these limits at points that will easily allow, for example, for unusually long but real content without allowing absurdly long, unreal content that might DOS the DB.

### Reproduction Steps

1. Create a new Jellyfin instance with a fresh `/config/` directory.
2. Include a library of media files with some `**/*.nfo` files containing `<title>` elements 100+ MB long.
3. Add that library to Jellyfin.
4. Note that the DB and `RES` RAM usage grow to unreasonable sizes.

### Jellyfin Version

10.9.0

### if other:

10.9.7

### Environment

- OS:

```console
# cat /etc/os-release
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os
```
- Linux Kernel:
```console
# uname -a
Linux library.moodysalonnet 5.15.153.1-microsoft-standard-WSL2 #1 SMP Fri Mar 29 23:14:13 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```
- Virtualization:
```console
# docker --version
Docker version 26.1.4, build 5650f9b
```
- Clients: Web and Kodi
- Browser: Google Chrome Version 128.0.6559.0 (Official Build) dev (64-bit)
- FFmpeg Version:
```console
# docker image ls --digests $(docker compose config --images jellyfin)
REPOSITORY                  TAG       DIGEST                                                                    IMAGE ID       CREATED       SIZE
ghcr.io/jellyfin/jellyfin   latest    sha256:76930362d2c266e8f0b2f51f1c6b039a5207525d0e138a1cdbec870c66c34b0f   49572434ce4c   12 days ago   1.04GB
```
- Playback Method: N/A
- Hardware Acceleration: N/A
- GPU Model: N/A
- Plugins: Kodi
- Reverse Proxy:
```console
# docker image ls --digests $(docker compose config --images traefik)
REPOSITORY   TAG       DIGEST                                                                    IMAGE ID       CREATED      SIZE
traefik      latest    sha256:a208c74fd80a566d4ea376053bff73d31616d7af3f1465a7747b8b89ee34d97e   7a4ed730cae1   4 days ago   172MB
```
- Base URL: `/`
- Networking: `network_mode: host`
- Storage: local, `/config/` on encrypted NVMe, libraries on Linux md RAID5T

### Jellyfin logs

N/A

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by rpatterson** [comment]
My first observed symptom was the slow startup causing `HEALTHCHECK` failures, so see [my comment there](https://github.com/jellyfin/jellyfin/issues/10225#issuecomment-2212522937) if you want more context.

**Comment by Zachwang007** [comment]
my is same.when the library scan, a huge ram on use, untill full ram used,and jellyfin stop

**Comment by cvium** [comment]
Gonna need to see those nfo files

**Comment by rpatterson** [comment]
> Gonna need to see those nfo files

Sorry, they're gone now, but also I wouldn't share them for privacy reasons. But I'd be happy to elaborate on the details [I included in the comment in another issue that I linked to in my comment](https://github.com/jellyfin/jellyfin/issues/12232#issuecomment-2212547047):

> Inspecting the worst offenders, I found that they contained extremely long `<title>` elements with tons of repeated strings separated by ` / `, including things that are definitely not parts of the title, such as actors' names.

These  `<title>` elements comprised ~99.99...% of the offending `**.nfo` file sizes. Let me know specifically what details you need and I'll reply with them if at all possible, @cvium.

**Comment by rpatterson** [comment]
> my is same.when the library scan, a huge ram on use, untill full ram used,and jellyfin stop

See [my comment in another issue that I linked to in my comment](https://github.com/jellyfin/jellyfin/issues/12232#issuecomment-2212547047) for workarounds, @Zachwang007.

**Comment by felix920506** [comment]
Yeah if you can't provide an NFO file that creates the issue then it's not going to be fixed.

If you don't want to share one from your library then I suggest you manufacture one with random data in a text editor or something and provide it.

**Comment by felix920506** [comment]
@OdinVex please open a new issue for your problem.

**Comment by rpatterson** [comment]
> If you don't want to share one from your library then I suggest you manufacture one with random data in a text editor or something and provide it.

From my steps to reproduce:

> Include a library of media files with some **/*.nfo files containing <title> elements 100+ MB long.

So you'd rather I **upload** a file with 100+ MB of repeated nonsense than provide the recipe for creating it? That seems like a waste of resources and doesn't make sense to me. I can't imagine the absence of the file when a recipe for creating was provided is the reason this issue isn't being investigated.

I imagine the reason it isn't being investigated is because contributor time is precious, constrained and not centrally directed in open source community projects, and because not enough other users know they're affected by this issue. I don't have enough experience with Jellyfin's stack to do it myself and, especially given that I have a workaround, I'm not going to direct my free time to learning its stack. Since I'm not going to do it, I don't get to complain about anyone else not doing it, those are the open source values I was taught. I still thought my findings of all the debugging time I did were worth capturing, especially the workarounds, in case it ever does affect more users, another open source value I was taught. This is exactly what closing an issue as `not planned` is for, or leave it open, up to the maintainer's preferred system.

But there's no need to justify closing an issue by saying the report is incomplete when the reporter put significant effort in to get it right. Ditto for asking the reporter to re-test every "n" months to keep it open. Both of these approaches discourage reporting and potential future contributors to no good benefit. If the maintainers prefer to manage issues by closing issues that won't be worked on, then just close them as `not planned` at your discretion. I would encourage maintainers to have their bots say something like what I've seen on other projects:

> Contributor time on open source community projects is constrained and this issue doesn't seem to have grabbed enough contributor interest to start work. To help the maintainers manage issues, we're closing this as `not planned`. If you or any other contributor wants to start work on this issue, we'll be happy to re-open it then.

IOW, issue closing bots can be both merciless, to not waste maintainers time, and kind, to foster community participation.

This issue management problem isn't specific to Jellyfin, it's all over open source GitHub projects, and Jellyfin is far from the worst I've encountered. Please forgive the OT rant. ;-)

**Comment by rpatterson** [comment]
Still waiting for follow up on the steps to reproduce.
